### PR TITLE
Joey dev

### PR DESCRIPTION
**Pull request by joeythegod**
Finish week 2 tutorial

### CODE REVIEW

**Comment by joeythegod** [comment]
I add the link to my deployed site~ Feel free to leave a message there XD

**Review by 0charleschen0 — COMMENTED** [review]
Could you put more contents on each of your commits?
Such as what this commit can achieve?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -19,14 +19,52 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;
+
+import com.google.gson.Gson;
+
+
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:32]: Add javaDoc here.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -19,14 +19,52 @@
...
+class Comment{
+    private String name;
+    private String content;
+    Comment(String name, String content){
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: a empty line here.

File: portfolio/src/main/webapp/index.html [hunk 3]
```diff
@@ -15,7 +15,7 @@
         <!-- Core theme CSS (includes Bootstrap)-->
         <link href="style.css" rel="stylesheet" />
     </head>
-    <body id="page-top" onload="getNameUsingFetch()">
+    <body id="page-top" onload="getNameUsingFetch();getComments()">
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/index.html]: nit: a space between getNameUsingFetch(); and getComment().

File: portfolio/src/main/webapp/index.html [hunk 4]
```diff
@@ -54,7 +54,8 @@ <h1 class="mb-0">
                         <a class="social-icon" href="https://www.facebook.com/joey19960318"><i class="fab fa-facebook-f"></i></a>
                         <a class="social-icon" href="https://www.kaggle.com/joeychuang"><i class="fab fa-kaggle"></i></a>
                     </div>
-                    <div id="hello-container"></div>
+                    <!-- <div id="hello-container"></div> -->
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/index.html]: Please remove this if don't needed

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -46,3 +46,23 @@ function getNameUsingFetch() {
     document.getElementById('hello-container').innerText = quote;
   });
 }
+
+function getComments() {
+    fetch('/data')  // sends a request to /my-data-url
+    .then(response => response.json()) // parses the response as JSON
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/script.js]: +4 ident for this line. L53 also

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -46,3 +46,23 @@ function getNameUsingFetch() {
...
+    .then((comments) => { // now we can reference the fields in myObject!
+    console.log(comments);
+    const commentsListElement = document.getElementById('comments-container');
+    commentsListElement.innerHTML = '';
+    for (let index in comments) {
+      commentsListElement.appendChild(createListElement(comments[index].name+' says '+comments[index].content));
+    }
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/script.js]: L54 - L59 +2 ident

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -19,14 +19,52 @@
...
+    comments.add(content2);
+    comments.add(content3);
+    String json = convertToJsonUsingGson(comments);
+
+    // Send the JSON as the response
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+    /**
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Please follow coding style.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -19,14 +19,52 @@
...
+   * Converts a ServerStats instance into a JSON string using the Gson library. Note: We first added
+   * the Gson library dependency to pom.xml.
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
   Note: We first added the Gson library dependency to pom.xml.
```
Not needed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -19,14 +19,52 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;
+
+import com.google.gson.Gson;
+
+
+class Comment{
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Consider to change to private if outer class doesn't need to access it.

**Review by 0charleschen0 — COMMENTED** [review]
Please file another pull request for commit 4 and 5

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -38,24 +38,35 @@
...
-    comments.add(content1);
-    comments.add(content2);
-    comments.add(content3);
+    // List<Comment> comments = new ArrayList<>();
+    // Comment content1 = new Comment("Jack","Hi I am Jack");
+    // Comment content2 = new Comment("John","Hi I am John");
+    // Comment content3 = new Comment("Jimmy","Hi I am Jimmy");
+    // comments.add(content1);
+    // comments.add(content2);
+    // comments.add(content3);
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Why comment this out? Please remove it if not needed anymore.

File: portfolio/src/main/webapp/index.html [hunk 11]
```diff
@@ -55,7 +56,7 @@ <h1 class="mb-0">
                         <a class="social-icon" href="https://www.kaggle.com/joeychuang"><i class="fab fa-kaggle"></i></a>
                     </div>
                     <!-- <div id="hello-container"></div> -->
-                    <ul id="comments-container"></ul>
+                    <!-- <ul id="comments-container"></ul> -->
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/index.html]: This one also

File: portfolio/src/main/webapp/index.html [hunk 12]
```diff
@@ -237,7 +238,23 @@ <h2>Taiko on NSwitch</h2>
...
+            </section>
+            <hr class="m-0" />
+            <!-- Message-->
+            <section class="resume-section" id="message">
+                <div class="resume-section-content">
+                    <h2 class="mb-5">Message me</h2>
+                    <ul id="messages-container"></ul>
+                    <form action="/data" method="POST">
+                        <p>Your name:</p>
+                        <textarea name="text-input-1"></textarea>
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/index.html]: Please use more self-explanatory naming such as "text-input-name", "text-input-content"

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -41,28 +41,40 @@
...
+//     .then(response => response.json()) // parses the response as JSON
+//     .then((comments) => { // now we can reference the fields in myObject!
+//     console.log(comments);
+//     const commentsListElement = document.getElementById('comments-container');
+//     commentsListElement.innerHTML = '';
+//     for (let index in comments) {
+//       commentsListElement.appendChild(createListElement(comments[index].name+' says '+comments[index].content));
+//     }
+//     });
+// }
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/script.js]: Why commenting out this content? Please remove if not necessary.

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -41,28 +41,40 @@
...
+
+/** Creates an <li> element containing text. */
+function createListElement(text) {
+  const liElement = document.createElement('li');
+  liElement.innerText = text;
+  return liElement;
 }
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/script.js:49]: Add JsDoc

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 15]
```diff
@@ -23,34 +23,42 @@
...
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 class Comment{
     private String name;
     private String content;
-    Comment(String name, String content){
+    private long timestamp;
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:44]: An empty line below.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 16]
```diff
@@ -23,34 +23,42 @@
...
-    // List<Comment> comments = new ArrayList<>();
-    // Comment content1 = new Comment("Jack","Hi I am Jack");
-    // Comment content2 = new Comment("John","Hi I am John");
-    // Comment content3 = new Comment("Jimmy","Hi I am Jimmy");
-    // comments.add(content1);
-    // comments.add(content2);
-    // comments.add(content3);
+    List<Comment> comments = new ArrayList<>();
+    Query query = new Query("comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: DatastoreService datastore appears multiple times. Could this be a member of DataServlet?

**Review by evalee20000831 — APPROVED** [review]

**Review by teefang — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (joeythegod) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Sorry I am not sure what to do about this comment.

File: portfolio/pom.xml [hunk 17]
```diff
@@ -23,7 +23,18 @@
...
     </dependency>
+    <dependency>
+        <groupId>com.google.code.gson</groupId>
+        <artifactId>gson</artifactId>
+        <version>2.8.6</version>
+    </dependency>
+    <dependency>
+        <groupId>com.google.appengine</groupId>
+        <artifactId>appengine-api-1.0-sdk</artifactId>
+        <version>1.9.59</version>
```
Comment (0charleschen0) [inline portfolio/pom.xml]: Idents
```suggestion
      <groupId>com.google.code.gson</groupId>
      <artifactId>gson</artifactId>
      <version>2.8.6</version>
    </dependency>
    <dependency>
      <groupId>com.google.appengine</groupId>
      <artifactId>appengine-api-1.0-sdk</artifactId>
      <version>1.9.59</version>
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 18]
```diff
@@ -19,14 +19,90 @@
...
+  /**
+   * Define the object of a comment for message section.
+   * @param name The author of the comment.
+   * @param content The content of the comment.
+   * @param timestamp Automatically record the timestamp of the comment submit.
+   */
+  private class Comment{
+    private String name;
+    private String content;
+    private long timestamp;
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: @param is only used to document method, not class.

```suggestion
  /** Define the object of a comment for message section. */
  private class Comment{
    /** The author of the comment */
    private String name;
    /** The content of the comment */
    private String content;
    /** The timestamp of the comment submission which records automatically */
    private long timestamp;
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 16 above)
Comment (0charleschen0) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Could you reply this comment?

File: portfolio/src/main/webapp/index.html (hunk 12 above)
Comment (0charleschen0) [reply portfolio/src/main/webapp/index.html]: This one also.

File: portfolio/src/main/webapp/script.js [hunk 19]
```diff
@@ -41,8 +41,34 @@
...
+//         .then(response => response.json()) // parses the response as JSON
+//         .then((comments) => { // now we can reference the fields in myObject!
+//             console.log(comments);
+//             const commentsListElement = document.getElementById('comments-container');
+//             commentsListElement.innerHTML = '';
+//             for (let index in comments) {
+//                 commentsListElement.appendChild(createListElement(comments[index].name+' says '+comments[index].content));
+//             }
+//         });
+// }
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/script.js]: Why comments this out?
Please remove it if not needed.

**Review by 0charleschen0 — COMMENTED** [review]
Looks better except some nitpicking.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (joeythegod) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: OK!

File: portfolio/src/main/webapp/script.js (hunk 14 above)
Comment (joeythegod) [reply portfolio/src/main/webapp/script.js:49]: In the latest video chat, you said this function is clear so it doesn't need to add JsDoc?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 16 above)
Comment (joeythegod) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Sure, I will add it as a member of DataServlet.

File: portfolio/src/main/webapp/script.js [hunk 20]
```diff
@@ -41,8 +36,25 @@
...
+/** 
+ * Creates an <li> element containing text.
+ * @param text The value of the element.
+ * @return The list-element.
+ */
+function createListElement(text) {
+    const liElement = document.createElement('li');
+    liElement.innerText = text;
+    return liElement;
 }
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/script.js]:
```suggestion
function createListElement(text) {
  const liElement = document.createElement('li');
  liElement.innerText = text;
  return liElement;
}
```

File: portfolio/src/main/webapp/script.js [hunk 21]
```diff
@@ -41,8 +36,25 @@
...
+    fetch('/data')  // sends a request to /my-data-url
+        .then(response => response.json()) // parses the response as JSON
+        .then((comments) => { // now we can reference the fields in myObject!
+            const commentsListElement = document.getElementById('messages-container');
+            commentsListElement.innerHTML = '';
+            for (let index in comments) {
+                commentsListElement.appendChild(createListElement(comments[index].name+' says '+comments[index].content));
+            }
+        });
+}
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/script.js]:
```suggestion
function getMessages() {
  fetch('/data')  // sends a request to /my-data-url
      .then(response => response.json()) // parses the response as JSON
      .then((comments) => { // now we can reference the fields in myObject!
        const commentsListElement = document.getElementById('messages-container');
        commentsListElement.innerHTML = '';
        for (let index in comments) {
          commentsListElement.appendChild(createListElement(comments[index].name+' says '+comments[index].content));
        }
      });
}
```

**Review by 0charleschen0 — COMMENTED** [review]
2 indent issues. Others LGTM.

**Pull request merged**